src/main/java/com/projectg/geyserupdater/common/util/FileUtils.java

```diff
@@ -58,33 +59,57 @@ public static boolean checkFile(String path, boolean allowCached) {
      * @param fileURL the url of the file
      * @param outputPath the path of the output file to write to
      */
-    public static void downloadFile(String fileURL, String outputPath) throws IOException {
-        // TODO: better download code?
+    public static void downloadFile(String fileURL, String outputPath, @NotNull Platform platform) throws IOException {
+        UpdaterLogger logger = UpdaterLogger.getLogger();
+        logger.debug("Attempting to download a file with URL and output path: " + fileURL + " , " + outputPath);
 
-        UpdaterLogger.getLogger().debug("Attempting to download a file with URL and output path: " + fileURL + " , " + outputPath);
+        // TODO: this whole cached thing only works if you're using checkFile for one file...
 
         Path outputDirectory = Paths.get(outputPath).getParent();
         Files.createDirectories(outputDirectory);
-
-        OutputStream os;
-        InputStream is;
-        // create a url object
+        // Download Jar file
         URL url = new URL(fileURL);
-        // connection to the file
-        URLConnection connection = url.openConnection();
-        // get input stream to the file
-        is = connection.getInputStream();
-        // get output stream to download file
-        os = new FileOutputStream(outputPath);
-        final byte[] b = new byte[2048];
-        int length;
-        // read from input stream and write to output stream
-        while ((length = is.read(b)) != -1) {
-            os.write(b, 0, length);
+        ReadableByteChannel rbc = Channels.newChannel(url.openStream());
+        FileOutputStream fos = new FileOutputStream(outputPath);
+        fos.getChannel().transferFrom(rbc, 0, Long.MAX_VALUE);
+        fos.close();
+        rbc.close();
+        // Checking file checksum
+        String sha256 = null;
+        switch (platform) {
+            case SPIGOT -> sha256 = new GeyserApi().endPoints().downloads.spigot.sha256;
+            case BUNGEECORD -> sha256 = new GeyserApi().endPoints().downloads.bungeecord.sha256;
+            case VELOCITY -> sha256 = new GeyserApi().endPoints().downloads.velocity.sha256;
+        }
```

Comment:
A `Supplier<String>` of the expected hash should probably be given to this method instead, to keep the method generic